core/templates/dev/head/pages/library/library.html

```diff
@@ -125,7 +126,8 @@ <h2 ng-class="{'active': activeGroupIndex === $index}" class="oppia-library-grou
                                              node-count="tile.node_count"
                                              category="tile.category"
                                              thumbnail-icon-url="tile.thumbnail_icon_url"
-                                             thumbnail-bg-color="tile.thumbnail_bg_color">
+                                             thumbnail-bg-color="tile.thumbnail_bg_color"
+                                             learner-dashboard-activity-ids="learnerDashboardActivityIds">
```

Comment:
I didn't quite understand this. I have created a learner playlist service but can't we pass the ids from here itself? In whatever page we want to show the icon we just need to fetch the ids using LearnerDashboardIdsBackendService and then pass this data as a field to the summary. So if you think about it, passing this info itself acts like a bool whether to display the icons or not. If we do want to display icons we do not pass this information.

And I think this is easily extendable and maintainable also. If we want to display the icons on some page in the future - just fetch the ids and pass this info to the summary. 

How does this sound?

Reply:
I think the answer is still no. Passing data through a long chain of functions is a bit of a code smell. If you added additional sub-directives (between the activity grid and the library, say) or changed the layout of the page or added some other thing, then all the intermediate layers would have to carry the data along. It's also a bit like a global variable that everyone can edit -- you now have this info stored in the service, and in the top-level controller, and in all the subdirectives, and all of these have to be in sync. A few iterations of this will turn the whole page into a big ball of mud as additional changes get made in the future.

The proposal instead is to do everything using the service as a reference point, so that we have the canonical version of the data stored in exactly one place. This also avoids a lot of passing the data around and also avoids outer controllers having to hold more state than needed. Better for one service to take care of all that (i.e. initial load, and subsequent modifications).

Now, you may be wondering, what's the point of the boolean then, and why is a boolean OK when the list of activity IDs is not so OK? Well, that's because the boolean is essentially a UI thing -- it indicates whether or not to display that info in the tile, regardless of whether we actually have the data or not. A "false" flag means we never display it (maybe this is done because of the particular design of the tile); a "true" flag means we can display it if we have the data. This is a direct result of the context in which the directive is placed, and is something that will vary with context, so it's reasonable to re-specify that flag each time because the context is different each time. But the data will always be the same for that particular user.

Reply:
This is done!